core/primitives/src/time.rs

```diff
@@ -1,155 +1,295 @@
-    fn drop(&mut self) {
-        Clock::reset();
+    #[cfg(test)]
+    mod tests {
+        // TODO(#5345) Add tests for mock
+        use super::*;
+
+        #[test]
+        fn test_mock() {
+            Clock::set_mock();
+            let utc = Utc::now();
+
+            Clock::add_utc(utc);
+
+            assert_eq!(Clock::utc(), utc);
+        }
     }
 }
 
-impl Clock {
-    pub fn set_mock() {
-        MockClockPerThread::with(|clock| {
-            clock.is_mock = true;
-        });
-    }
-    pub fn reset() {
-        MockClockPerThread::with(|clock| {
-            clock.reset();
-        });
+mod time {
+    use crate::time::{Clock, Duration, Utc};
+    use borsh::{BorshDeserialize, BorshSerialize};
+    use chrono::DateTime;
+    use std::ops::{Add, Sub};
+    use std::time::SystemTime;
+
+    pub const TIME_UNIX_EPOCH: Time = Time::from_system_time(SystemTime::UNIX_EPOCH);
+
+    #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+    pub struct Time {
+        time: SystemTime,
     }
-    pub fn add_utc(mock_date: DateTime<chrono::Utc>) {
-        MockClockPerThread::with(|clock| {
-            if clock.is_mock {
-                clock.utc.push_back(mock_date);
-            } else {
-                panic!("Use MockClockGuard in your test");
-            }
-        });
+
+    impl BorshSerialize for Time {
+        fn serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<(), std::io::Error> {
+            let nanos = self.duration_since(&TIME_UNIX_EPOCH).as_nanos() as u64;
+            BorshSerialize::serialize(&nanos, writer).unwrap();
+            Ok(())
+        }
     }
 
-    pub fn add_instant(mock_instant: Duration) {
-        MockClockPerThread::with(|clock| {
-            if clock.is_mock {
-                clock.durations.push_back(mock_instant);
-            } else {
-                panic!("Use MockClockGuard in your test");
-            }
-        });
+    impl BorshDeserialize for Time {
+        fn deserialize(buf: &mut &[u8]) -> Result<Self, std::io::Error> {
+            let nanos: u64 = borsh::BorshDeserialize::deserialize(buf)?;
+
+            Ok(TIME_UNIX_EPOCH + Duration::from_nanos(nanos))
+        }
     }
 
-    pub fn utc() -> DateTime<chrono::Utc> {
-        MockClockPerThread::with(|clock| {
-            if clock.is_mock {
-                let x = clock.pop_utc();
-                match x {
-                    Some(t) => t,
-                    None => {
-                        panic!("Mock clock run out of samples");
-                    }
-                }
-            } else {
-                chrono::Utc::now()
-            }
-        })
+    impl Time {
+        pub const fn from_system_time(system_time: SystemTime) -> Time {
+            Time { time: system_time }
+        }
+
+        pub fn from_utc(utc: DateTime<Utc>) -> Time {
+            let nanos = utc.timestamp_nanos() as u64;
+
+            TIME_UNIX_EPOCH + Duration::from_nanos(nanos)
+        }
```

Comment:
These two can alternatively be `From` impls.

Reply:
Thanks, done.